File: src/dbus/ServiceExplorer.cpp
#include "ServiceExplorer.h"
#include "DBusManager.h"

#include <QDBusInterface>
#include <QDBusReply>
#include <QDomDocument>
#include <QDebug>
#include <QVariantMap>

ServiceExplorer::ServiceExplorer(QObject *parent)
    : QObject(parent)
{
}

QStringList ServiceExplorer::listServices(int busType)
{
    QStringList services;

    QDBusConnection conn = (static_cast<DBusManager::BusType>(busType) == DBusManager::SessionBus) ? QDBusConnection::sessionBus() : QDBusConnection::systemBus();
    if (!conn.isConnected()) {
        qWarning() << "[ServiceExplorer] Not connected to D-Bus busType=" << busType;
        return services;
    }

    QDBusInterface dbus("org.freedesktop.DBus", "/org/freedesktop/DBus", "org.freedesktop.DBus", conn);
    QDBusReply<QStringList> reply = dbus.call("ListNames");

    if (reply.isValid()) {
        services = reply.value();
        qDebug() << "[ServiceExplorer] listServices returned" << services.count() << "names";
    } else {
        qWarning() << "[ServiceExplorer] Failed to list services:" << reply.error().message();
    }

    return services;
}

DbusServiceInfo ServiceExplorer::introspectService(const QString &serviceName, int busType)
{
    DbusServiceInfo info;
    info.name = serviceName;
    QString id = serviceName;
    info.id = id.replace(".", "-").replace(":", "");
    info.busType = static_cast<DBusManager::BusType>(busType);

    QDBusConnection conn = (static_cast<DBusManager::BusType>(busType) == DBusManager::SessionBus) ? QDBusConnection::sessionBus() : QDBusConnection::systemBus();
    if (!conn.isConnected()) {
        qWarning() << "Not connected to D-Bus";
        return info;
    }

    const QString rootPath = "/";
    info.interfaces = introspectPath(serviceName, rootPath, busType);

    emit introspectionCompleted(info);

    return info;
}

QVariantList ServiceExplorer::introspectServiceMembers(const QString &serviceName, int busType)
{
    QVariantList members;
    const DbusServiceInfo serviceInfo = introspectService(serviceName, busType);

    for (const DbusInterfaceInfo &interfaceInfo : serviceInfo.interfaces) {
        for (const DbusMemberInfo &member : interfaceInfo.methods) {
            members.append(m_memberToVariantMap(member));
        }
        for (const DbusMemberInfo &member : interfaceInfo.signalMembers) {
            members.append(m_memberToVariantMap(member));
        }
        for (const DbusMemberInfo &member : interfaceInfo.properties) {
            members.append(m_memberToVariantMap(member));
        }
    }

    return members;
}

QList<DbusInterfaceInfo> ServiceExplorer::introspectPath(const QString &serviceName, const QString &path, int busType)
{
    QList<DbusInterfaceInfo> interfaces;

    QString xml = m_introspect(serviceName, path, busType);
    if (xml.isEmpty()) {
        return interfaces;
    }

    QDomDocument doc;
    if (!doc.setContent(xml)) {
        qWarning() << "Failed to parse introspection XML for" << serviceName << path;
        return interfaces;
    }

    QDomElement root = doc.documentElement();

    for (QDomNode child = root.firstChild(); !child.isNull(); child = child.nextSibling()) {
        const QDomElement childElement = child.toElement();
        if (childElement.isNull()) {
            continue;
        }

        if (childElement.tagName() == "interface") {
            const QString interfaceName = childElement.attribute("name");
            if (interfaceName.startsWith("org.freedesktop.DBus.")) {
                continue;
            }

            DbusInterfaceInfo interfaceInfo;
            interfaceInfo.name = interfaceName;
            interfaceInfo.path = path;
            interfaceInfo.methods = m_parseMethods(const_cast<QDomElement &>(childElement), path);
            interfaceInfo.signalMembers = m_parseSignals(const_cast<QDomElement &>(childElement), path);
            interfaceInfo.properties = m_parseProperties(const_cast<QDomElement &>(childElement), path);
            interfaces.append(interfaceInfo);
            continue;
        }

        if (childElement.tagName() != "node") {
            continue;
        }

        const QString subPath = childElement.attribute("name");
        if (subPath.isEmpty()) {
            continue;
        }

        const QString fullPath = path == "/" ? "/" + subPath : path + "/" + subPath;
        interfaces.append(introspectPath(serviceName, fullPath, busType));
    }

    return interfaces;
}

QString ServiceExplorer::m_introspect(const QString &serviceName, const QString &path, int busType)
{
    QDBusConnection conn = (static_cast<DBusManager::BusType>(busType) == DBusManager::SessionBus) ? QDBusConnection::sessionBus() : QDBusConnection::systemBus();
    if (!conn.isConnected()) {
        return QString();
    }

    QDBusInterface iface(serviceName, path, "org.freedesktop.DBus.Introspectable", conn);
    QDBusReply<QString> reply = iface.call("Introspect");

    if (reply.isValid()) {
        return reply.value();
    } else {
        qWarning() << "Introspection failed for" << serviceName << path << ":" << reply.error().message();
        emit introspectionFailed(serviceName, reply.error().message());
        return QString();
    }
}

bool ServiceExplorer::m_parseIntrospectionXml(const QString &xml, DbusInterfaceInfo &interfaceInfo)
{
    QDomDocument doc;
    if (!doc.setContent(xml)) {
        return false;
    }

    QDomElement root = doc.documentElement();

    QDomNodeList interfaceNodes = root.elementsByTagName("interface");
    for (int i = 0; i < interfaceNodes.count(); ++i) {
        QDomElement interfaceElement = interfaceNodes.item(i).toElement();
        if (interfaceElement.attribute("name") == interfaceInfo.name) {
            interfaceInfo.methods = m_parseMethods(interfaceElement, interfaceInfo.path);
            interfaceInfo.signalMembers = m_parseSignals(interfaceElement, interfaceInfo.path);
            interfaceInfo.properties = m_parseProperties(interfaceElement, interfaceInfo.path);
            return true;
        }
    }

    return false;
}

QList<DbusMemberInfo> ServiceExplorer::m_parseMethods(QDomElement &interfaceElement, const QString &path)
{
    QList<DbusMemberInfo> methods;
    QDomNodeList methodNodes = interfaceElement.elementsByTagName("method");

    for (int i = 0; i < methodNodes.count(); ++i) {
        QDomElement methodElement = methodNodes.item(i).toElement();
        DbusMemberInfo info;
        info.id = path + "|" + interfaceElement.attribute("name") + "|method|" + methodElement.attribute("name");
        info.name = methodElement.attribute("name");
        info.type = "method";
        info.interfaceName = interfaceElement.attribute("name");
        info.path = path;

        QDomNodeList argNodes = methodElement.elementsByTagName("arg");
        for (int j = 0; j < argNodes.count(); ++j) {
            QDomElement argElement = argNodes.item(j).toElement();
            QString direction = argElement.attribute("direction", "in");
            if (direction == "in") {
                info.signature += argElement.attribute("type");
            } else {
                info.returnType += argElement.attribute("type");
            }
        }

        QDomNodeList annotationNodes = methodElement.elementsByTagName("annotation");
        for (int k = 0; k < annotationNodes.count(); ++k) {
            QDomElement annotationElement = annotationNodes.item(k).toElement();
            if (annotationElement.attribute("name") == "org.freedesktop.DBus.Deprecated") {
                info.annotation = "deprecated";
            }
        }

        methods.append(info);
    }

    return methods;
}

QList<DbusMemberInfo> ServiceExplorer::m_parseSignals(QDomElement &interfaceElement, const QString &path)
{
    QList<DbusMemberInfo> result;
    QDomNodeList signalNodes = interfaceElement.elementsByTagName("signal");

    for (int i = 0; i < signalNodes.count(); ++i) {
        QDomElement signalElement = signalNodes.item(i).toElement();
        DbusMemberInfo info;
        info.id = path + "|" + interfaceElement.attribute("name") + "|signal|" + signalElement.attribute("name");
        info.name = signalElement.attribute("name");
        info.type = "signal";
        info.interfaceName = interfaceElement.attribute("name");
        info.path = path;

        QDomNodeList argNodes = signalElement.elementsByTagName("arg");
        for (int j = 0; j < argNodes.count(); ++j) {
            QDomElement argElement = argNodes.item(j).toElement();
            info.signature += argElement.attribute("type");
        }

        result.append(info);
    }

    return result;
}

QList<DbusMemberInfo> ServiceExplorer::m_parseProperties(QDomElement &interfaceElement, const QString &path)
{
    QList<DbusMemberInfo> properties;
    QDomNodeList propertyNodes = interfaceElement.elementsByTagName("property");

    for (int i = 0; i < propertyNodes.count(); ++i) {
        QDomElement propertyElement = propertyNodes.item(i).toElement();
        DbusMemberInfo info;
        info.id = path + "|" + interfaceElement.attribute("name") + "|property|" + propertyElement.attribute("name");
        info.name = propertyElement.attribute("name");
        info.type = "property";
        info.interfaceName = interfaceElement.attribute("name");
        info.path = path;
        info.signature = propertyElement.attribute("type");

        QString access = propertyElement.attribute("access", "read");
        if (access.contains("read")) {
            if (access.contains("write")) {
                info.annotation = "readwrite";
            } else {
                info.annotation = "readonly";
            }
        } else {
            info.annotation = "writeonly";
        }

        properties.append(info);
    }

    return properties;
}

QVariantMap ServiceExplorer::m_memberToVariantMap(const DbusMemberInfo &member) const
{
    return {
        {"id", member.id},
        {"name", member.name},
        {"type", member.type},
        {"interfaceName", member.interfaceName},
        {"path", member.path},
        {"signature", member.signature},
        {"returnType", member.returnType},
        {"annotation", member.annotation}
    };
}

File: electron/dbus/ServiceExplorer.ts
import { Message, MessageType, sessionBus, systemBus } from 'dbus-next'
import { readFileSync } from 'fs'
import type { BusType, DbusMemberInfo, DbusInterfaceInfo, DbusArgumentInfo, ServiceInfo } from './types'

/**
 * ServiceExplorer - Discover and introspect D-Bus services
 *
 * This class provides functionality to:
 * - List all services on a D-Bus bus
 * - Introspect services to discover interfaces, methods, signals, and properties
 * - Parse introspection XML into structured TypeScript objects
 */
export class ServiceExplorer {
  /**
   * List all services on the specified bus
   */
  async listServices(busType: BusType): Promise<string[]> {
    const bus = this.getBus(busType)

    try {
      // Call org.freedesktop.DBus.ListNames
      const method = new Message({
        type: MessageType.METHOD_CALL,
        destination: 'org.freedesktop.DBus',
        path: '/org/freedesktop/DBus',
        interface: 'org.freedesktop.DBus',
        member: 'ListNames',
      })

      const reply = await bus.call(method)

      if (reply.type !== MessageType.METHOD_RETURN) {
        throw new Error('Failed to list services')
      }

      const names: string[] = reply.body[0]

      // Filter out unique names (starting with ':') and sort
      const services = names
        .filter(name => !name.startsWith(':'))
        .sort()

      return services
    } finally {
      bus.disconnect()
    }
  }

  /**
   * Get detailed info about a service: unique name, PID, process command
   */
  async getServiceInfo(serviceName: string, busType: BusType): Promise<ServiceInfo> {
    const bus = this.getBus(busType)

    try {
      // Get unique name via GetNameOwner
      let uniqueName: string | null = null
      try {
        const nameOwnerReply = await bus.call(new Message({
          type: MessageType.METHOD_CALL,
          destination: 'org.freedesktop.DBus',
          path: '/org/freedesktop/DBus',
          interface: 'org.freedesktop.DBus',
          member: 'GetNameOwner',
          signature: 's',
          body: [serviceName],
        }))
        if (nameOwnerReply.type === MessageType.METHOD_RETURN) {
          uniqueName = nameOwnerReply.body[0]
        }
      } catch {
        // Service not active
      }

      if (!uniqueName) {
        return { serviceName, uniqueName: null, pid: null, processCmd: null, isActive: false }
      }

      // Get PID via GetConnectionUnixProcessID
      let pid: number | null = null
      try {
        const pidReply = await bus.call(new Message({
          type: MessageType.METHOD_CALL,
          destination: 'org.freedesktop.DBus',
          path: '/org/freedesktop/DBus',
          interface: 'org.freedesktop.DBus',
          member: 'GetConnectionUnixProcessID',
          signature: 's',
          body: [uniqueName],
        }))
        if (pidReply.type === MessageType.METHOD_RETURN) {
          pid = pidReply.body[0]
        }
      } catch {
        // PID not available
      }

      // Read process command from /proc/{pid}/cmdline
      let processCmd: string | null = null
      if (pid) {
        try {
          const cmdline = readFileSync(`/proc/${pid}/cmdline`, 'utf-8')
          // cmdline is null-separated, convert to space-separated
          processCmd = cmdline.split('\0').filter(Boolean).join(' ')
        } catch {
          // Process may have exited or /proc not available
        }
      }

      return { serviceName, uniqueName, pid, processCmd, isActive: true }
    } finally {
      bus.disconnect()
    }
  }

  /**
   * Introspect a service and return flat member list
   * This matches the Qt ServiceExplorer::introspectServiceMembers API
   */
  async introspectServiceMembers(
    serviceName: string,
    busType: BusType
  ): Promise<DbusMemberInfo[]> {
    const bus = this.getBus(busType)
    const members: DbusMemberInfo[] = []

    try {
      // Start with root path
      const paths = await this.explorePaths(bus, serviceName, '/')

      // Introspect each path
      for (const path of paths) {
        const interfaces = await this.introspectPath(bus, serviceName, path)

        // Flatten all members into a single list
        for (const iface of interfaces) {
          // Add methods
          for (const method of iface.methods) {
            members.push(method)
          }

          // Add signals
          for (const signal of iface.signalMembers) {
            members.push(signal)
          }

          // Add properties
          for (const property of iface.properties) {
            members.push(property)
          }
        }
      }

      return members
    } finally {
      bus.disconnect()
    }
  }

  /**
   * Recursively explore paths in a service
   */
  private async explorePaths(
    bus: any,
    serviceName: string,
    path: string
  ): Promise<string[]> {
    try {
      const paths: string[] = [path]
      const xml = await this.getIntrospectionXML(bus, serviceName, path)
      const nodeMatches = xml.match(/<node\s+name="([^"]+)"/g)

      if (nodeMatches) {
        for (const match of nodeMatches) {
          const nameMatch = match.match(/name="([^"]+)"/)
          if (nameMatch && nameMatch[1]) {
            const nodeName = nameMatch[1]
            const childPath = path === '/' ? `/${nodeName}` : `${path}/${nodeName}`
            const childPaths = await this.explorePaths(bus, serviceName, childPath)
            paths.push(...childPaths)
          }
        }
      }

      return paths
    } catch (error) {
      const isUnknownObject = (error as { type?: string }).type === 'org.freedesktop.DBus.Error.UnknownObject'
      if (isUnknownObject && path !== '/') {
        return []
      }

      console.error(`Failed to introspect path ${path}:`, error)
      return [path]
    }
  }

  /**
   * Introspect a specific path in a service
   */
  private async introspectPath(
    bus: any,
    serviceName: string,
    path: string
  ): Promise<DbusInterfaceInfo[]> {
    const interfaces: DbusInterfaceInfo[] = []

    try {
      const xml = await this.getIntrospectionXML(bus, serviceName, path)

      // Parse interfaces using simple regex
      const interfaceMatches = xml.match(/<interface\s+name="([^"]+)">([\s\S]*?)<\/interface>/g)
      if (!interfaceMatches) return interfaces

      for (const interfaceMatch of interfaceMatches) {
        const nameMatch = interfaceMatch.match(/name="([^"]+)"/)
        if (!nameMatch) continue

        const interfaceName = nameMatch[1]
        const iface: DbusInterfaceInfo = {
          name: interfaceName,
          path,
          methods: [],
          signalMembers: [],
          properties: [],
        }

        // Parse methods
        const methodMatches = interfaceMatch.match(/<method\s+name="([^"]+)">([\s\S]*?)<\/method>/g)
        if (methodMatches) {
          for (const methodMatch of methodMatches) {
            const methodNameMatch = methodMatch.match(/name="([^"]+)"/)
            if (!methodNameMatch) continue

            const methodName = methodNameMatch[1]
            const args = this.extractArguments(methodMatch, 'in')
            const inputArgs = args.filter((arg) => arg.direction === 'in')
            const outputArgs = args.filter((arg) => arg.direction === 'out')

            const method: DbusMemberInfo = {
              id: `${path}:${interfaceName}:${methodName}`,
              name: methodName,
              type: 'method',
              serviceName,
              interfaceName,
              path,
              signature: inputArgs.map((arg) => arg.type).join(''),
              returnType: outputArgs.map((arg) => arg.type).join(''),
              annotation: '',
              inputArgs,
              outputArgs,
            }

            iface.methods.push(method)
          }
        }

        // Parse signals
        const signalMatches = interfaceMatch.match(/<signal\s+name="([^"]+)">([\s\S]*?)<\/signal>/g)
        if (signalMatches) {
          for (const signalMatch of signalMatches) {
            const signalNameMatch = signalMatch.match(/name="([^"]+)"/)
            if (!signalNameMatch) continue

            const signalName = signalNameMatch[1]
            const outputArgs = this.extractArguments(signalMatch, 'out')

            const signal: DbusMemberInfo = {
              id: `${path}:${interfaceName}:${signalName}`,
              name: signalName,
              type: 'signal',
              serviceName,
              interfaceName,
              path,
              signature: outputArgs.map((arg) => arg.type).join(''),
              returnType: '',
              annotation: '',
              inputArgs: [],
              outputArgs,
            }

            iface.signalMembers.push(signal)
          }
        }

        // Parse properties
        const propertyMatches = interfaceMatch.match(/<property\s+name="([^"]+)"\s+type="([^"]+)"[^/]*\/>/g)
        if (propertyMatches) {
          for (const propertyMatch of propertyMatches) {
            const propertyNameMatch = propertyMatch.match(/name="([^"]+)"/)
            const propertyTypeMatch = propertyMatch.match(/type="([^"]+)"/)
            if (!propertyNameMatch || !propertyTypeMatch) continue

            const propertyName = propertyNameMatch[1]
            const propertyType = propertyTypeMatch[1]
            const accessMatch = propertyMatch.match(/access="([^"]+)"/)

            const property: DbusMemberInfo = {
              id: `${path}:${interfaceName}:${propertyName}`,
              name: propertyName,
              type: 'property',
              serviceName,
              interfaceName,
              path,
              signature: propertyType,
              returnType: propertyType,
              annotation: accessMatch ? accessMatch[1] : '',
              inputArgs: [],
              outputArgs: [],
            }

            iface.properties.push(property)
          }
        }

        interfaces.push(iface)
      }
    } catch (error) {
      console.error(`Failed to introspect path ${path}:`, error)
    }

    return interfaces
  }

  /**
   * Get introspection XML for a path
   */
  private async getIntrospectionXML(
    bus: any,
    serviceName: string,
    path: string
  ): Promise<string> {
    const method = new Message({
      type: MessageType.METHOD_CALL,
      destination: serviceName,
      path,
      interface: 'org.freedesktop.DBus.Introspectable',
      member: 'Introspect',
    })

    const reply = await bus.call(method)

    if (reply.type !== MessageType.METHOD_RETURN) {
      throw new Error(`Failed to introspect ${serviceName} at ${path}`)
    }

    return reply.body[0]
  }

  /**
   * Extract all argument nodes from XML as structured DbusArgumentInfo objects
   */
  private extractArguments(xml: string, defaultDirection: 'in' | 'out'): DbusArgumentInfo[] {
    const argMatches = xml.match(/<arg\b[^>]*\/?>/g)
    if (!argMatches) return []

    return argMatches.map((argXml) => ({
      name: this.extractAttribute(argXml, 'name'),
      type: this.extractAttribute(argXml, 'type'),
      direction: (this.extractAttribute(argXml, 'direction') || defaultDirection) as 'in' | 'out',
    }))
  }

  /**
   * Extract a single attribute value from an XML string
   */
  private extractAttribute(xml: string, name: string): string {
    const match = xml.match(new RegExp(`${name}="([^"]*)"`))
    return match?.[1] ?? ''
  }

  /**
   * Get D-Bus bus connection
   */
  private getBus(busType: BusType): any {
    if (busType === 'system') {
      return systemBus()
    } else {
      return sessionBus()
    }
  }
}
